pymc3/tests/test_csv_backend.py

```diff
@@ -0,0 +1,114 @@
+import numpy.testing as npt
+from pymc3.tests import backend_fixtures as bf
+from pymc3.backends import ndarray, csv
+
+
+class TestCSV0dSampling(bf.SamplingTestCase):
+    backend = csv.CSV
+    name = 'csv-db'
+    shape = ()
+
+
+class TestCSV1dSampling(bf.SamplingTestCase):
+    backend = csv.CSV
+    name = 'csv-db'
+    shape = 2
+
+
+class TestCSV2dSampling(bf.SamplingTestCase):
+    backend = csv.CSV
+    name = 'csv-db'
+    shape = (2, 3)
+
+
+class TestCSV0dSelection(bf.SelectionNoSliceTestCase):
+    backend = csv.CSV
+    name = 'csv-db'
+    shape = ()
+
+
+class TestCSV1dSelection(bf.SelectionNoSliceTestCase):
+    backend = csv.CSV
+    name = 'csv-db'
+    shape = 2
+
+
+class TestCSV2dSelection(bf.SelectionNoSliceTestCase):
+    backend = csv.CSV
+    name = 'csv-db'
+    shape = (2, 3)
+
+
+class TestCSVDumpLoad(bf.DumpLoadTestCase):
+    backend = csv.CSV
+    load_func = staticmethod(csv.load)
+    name = 'csv-db'
+    shape = (2, 3)
+
+
+class TestCSVDumpFunction(bf.BackendEqualityTestCase):
+    backend0 = backend1 = ndarray.NDArray
+    name0 = None
+    name1 = 'csv-db'
+    shape = (2, 3)
+
+    @classmethod
+    def setUpClass(cls):
+        super(TestCSVDumpFunction, cls).setUpClass()
+        csv.dump(cls.name1, cls.mtrace1)
+        with cls.model:
+            cls.mtrace1 = csv.load(cls.name1)
+
+
+class TestTraceToDf(bf.ModelBackendSampledTestCase):
+    backend = ndarray.NDArray
+    name = 'csv-db'
+    shape = (2, 3)
+
+    def test_trace_to_df(self):
+        mtrace = self.mtrace
+        df = csv._trace_to_df(mtrace._traces[0])
+        self.assertEqual(len(mtrace), df.shape[0])
+
+        xarr = mtrace.get_values('x', chains=0)
+        npt.assert_equal(xarr[:, 0, 0], df['x__0_0'].values)
+        npt.assert_equal(xarr[:, 1, 0], df['x__1_0'].values)
+        npt.assert_equal(xarr[:, 1, 2], df['x__1_2'].values)
+
+
+class TestNDArrayCSVEquality(bf.BackendEqualityTestCase):
+    backend0 = ndarray.NDArray
+    name0 = None
+    backend1 = csv.CSV
+    name1 = 'test.db'
+    shape = (2, 3)
+
+
+def test_create_flat_names_0d():
+    shape = ()
+    result = csv._create_flat_names('x', shape)
+    expected = ['x']
+    assert result == expected
+    assert csv._create_shape(result) == shape
+
+
+def test_create_flat_names_1d():
+    shape = 2,
+    result = csv._create_flat_names('x', shape)
+    expected = ['x__0', 'x__1']
```

Comment:
Just to make the indices visually distinct since it's common for
variable names to have both underscores and digits.  But if people
have a strong preference against it, I'm ok changing it to one
underscore.